README.md
# wct-james/gronx
Forked from [adhocore/gronx](https://github.com/adhocore/gronx)

[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Go Report](https://goreportcard.com/badge/github.com/wct-james/gronx)](https://goreportcard.com/report/github.com/wct-james/gronx)
[![Test](https://github.com/wct-james/gronx/actions/workflows/tests.yml/badge.svg)](https://github.com/wct-james/gronx/actions/workflows/tests.yml)
[![Codecov](https://img.shields.io/codecov/c/github/wct-james/gronx/main.svg?style=flat-square)](https://codecov.io/gh/wct-james/gronx)

> The reason for this fork is to make the implementation granular to minutes, rather than seconds. 
> 
> This is so that the package aligns more closely to the [Posix crontab specification](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/crontab.html).
> Whilst years remains, the `isDue()` method now returns true at any point in a due minute, rather than at the due second. 

`gronx` is Golang [cron expression](#cron-expression) parser ported from [adhocore/cron-expr](https://github.com/adhocore/php-cron-expr). You can also use gronx to find the next (`NextTick()`) or previous (`PrevTick()`) run time of an expression from any arbitrary point of time.

- Zero dependency.
- Very **fast** because it bails early in case a segment doesn't match.


Find gronx in [pkg.go.dev](https://pkg.go.dev/github.com/wct-james/gronx).

## Installation

```sh
go get -u github.com/wct-james/gronx
```

## Usage

```go
import (
	"time"

	"github.com/wct-james/gronx"
)

gron := gronx.New()
expr := "* * * * *"

// check if expr is even valid, returns bool
gron.IsValid(expr) // true

// check if expr is due for current time, returns bool and error
gron.IsDue(expr) // true|false, nil

// check if expr is due for given time
gron.IsDue(expr, time.Date(2021, time.April, 1, 1, 1, 0, 0, time.UTC)) // true|false, nil
```

### Batch Due Check

If you have multiple cron expressions to check due on same reference time use `BatchDue()`:
```go
gron := gronx.New()
exprs := []string{"* * * * *", "*/5 * * * *"}

// gives []gronx.Expr{} array, each item has Due flag and Err enountered.
dues := gron.BatchDue(exprs)

for _, expr := range dues {
    if expr.Err != nil {
        // Handle err
    } else if expr.Due {
        // Handle due
    }
}

// Or with given time
ref := time.Now()
gron.BatchDue(exprs, ref)
```

### Next Tick

To find out when is the cron due next (in near future):
```go
allowCurrent = true // includes current time as well
nextTime, err := gron.NextTick(expr, allowCurrent) // gives time.Time, error

// OR, next tick after certain reference time
refTime = time.Date(2022, time.November, 1, 1, 1, 0, 0, time.UTC)
allowCurrent = false // excludes the ref time
nextTime, err := gron.NextTickAfter(expr, refTime, allowCurrent) // gives time.Time, error
```

### Prev Tick

To find out when was the cron due previously (in near past):
```go
allowCurrent = true // includes current time as well
prevTime, err := gron.PrevTick(expr, allowCurrent) // gives time.Time, error

// OR, prev tick before certain reference time
refTime = time.Date(2022, time.November, 1, 1, 1, 0, 0, time.UTC)
allowCurrent = false // excludes the ref time
nextTime, err := gron.PrevTickBefore(expr, refTime, allowCurrent) // gives time.Time, error
```

> The working of `PrevTick*()` and `NextTick*()` are mostly the same except the direction.
> They differ in lookback or lookahead.

### Standalone Daemon

In a more practical level, you would use this tool to manage and invoke jobs in app itself and not
mess around with `crontab` for each and every new tasks/jobs.

In crontab just put one entry with `* * * * *` which points to your Go entry point that uses this tool.
Then in that entry point you would invoke different tasks if the corresponding Cron expr is due.
Simple map structure would work for this.

---
### Cron Expression

Cron expression is most commonly made of 5 segments. These segments are interpreted as:
```
<minute> <hour> <day> <month> <weekday>
```

In a 6 segments expression, if 6th segment matches `<year>` (i.e 4 digits at least) it will be interpreted as:
```
<minute> <hour> <day> <month> <weekday> <year>
```

For each segments you can have **multiple choices** separated by comma:
> Eg: `0,30 * * * *` means either 0th or 30th minute.

To specify **range of values** you can use dash:
> Eg: `10-15 * * * *` means 10th, 11th, 12th, 13th, 14th and 15th minute.

To specify **range of step** you can combine a dash and slash:
> Eg: `10-15/2 * * * *` means every 2 minutes between 10 and 15 i.e 10th, 12th and 14th minute.

For the `<day>` and `<weekday>` segment, there are additional [**modifiers**](#modifiers) (optional).

And if you want, you can mix the multiple choices, ranges and steps in a single expression:
> `5,12-20/4,55 * * * *` matches if any one of `5` or `12-20/4` or `55` matches the minute.

### Real Abbreviations

You can use real abbreviations (3 chars) for month and week days. eg: `JAN`, `dec`, `fri`, `SUN`

### Tags

Following tags are available and they are converted to real cron expressions before parsing:

- *@yearly* or *@annually* - every year
- *@monthly* - every month
- *@daily* - every day
- *@weekly* - every week
- *@hourly* - every hour
- *@5minutes* - every 5 minutes
- *@10minutes* - every 10 minutes
- *@15minutes* - every 15 minutes
- *@30minutes* - every 30 minutes
- *@always* - every minute

```go
// Use tags like so:
gron.IsDue("@hourly")
gron.IsDue("@5minutes")
```

### Modifiers

Following modifiers supported

- *Day of Month / 3rd of 5 segments / 4th of 6+ segments:*
    - `L` stands for last day of month (eg: `L` could mean 29th for February in leap year)
    - `W` stands for closest week day (eg: `10W` is closest week days (MON-FRI) to 10th date)
- *Day of Week / 5th of 5 segments / 6th of 6+ segments:*
    - `L` stands for last weekday of month (eg: `2L` is last monday)
    - `#` stands for nth day of week in the month (eg: `1#2` is second sunday)

---
## License

> &copy; [MIT](./LICENSE)

## Credits

This project is forked from [adhocore/gronx](https://github.com/adhocore/gronx)


CHANGELOG.md
## [v0.2.7](https://github.com/wct-james/gronx/releases/tag/v0.2.7) (2022-06-28)

### Miscellaneous
- **Workflow**: Run tests on 1.18x (Jitendra)
- Tests for go v1.17.x, add codecov (Jitendra)


## [v0.2.6](https://github.com/wct-james/gronx/releases/tag/v0.2.6) (2021-10-14)

### Miscellaneous
- Fix 'with' languages (Jitendra Adhikari) [_a813b55_](https://github.com/wct-james/gronx/commit/a813b55)
- Init/setup github codeql (Jitendra Adhikari) [_fe2aa5a_](https://github.com/wct-james/gronx/commit/fe2aa5a)


## [v0.2.5](https://github.com/wct-james/gronx/releases/tag/v0.2.5) (2021-07-25)

### Bug Fixes
- **Tasker**: The clause should be using OR (Jitendra Adhikari) [_b813b85_](https://github.com/wct-james/gronx/commit/b813b85)


## [v0.2.4](https://github.com/wct-james/gronx/releases/tag/v0.2.4) (2021-05-05)

### Features
- **Pkg.tasker**: Capture cmd output in tasker logger, error in stderr (Jitendra Adhikari) [_0da0aae_](https://github.com/wct-james/gronx/commit/0da0aae)

### Internal Refactors
- **Cmd.tasker**: Taskify is now method of tasker (Jitendra Adhikari) [_8b1373b_](https://github.com/wct-james/gronx/commit/8b1373b)


## [v0.2.3](https://github.com/wct-james/gronx/releases/tag/v0.2.3) (2021-05-04)

### Bug Fixes
- **Pkg.tasker**: Sleep 100ms so abort can be bailed asap, remove dup msg (Jitendra Adhikari) [_d868920_](https://github.com/wct-james/gronx/commit/d868920)

### Miscellaneous
- Allow leeway period at the end (Jitendra Adhikari) [_5ebf923_](https://github.com/wct-james/gronx/commit/5ebf923)


## [v0.2.2](https://github.com/wct-james/gronx/releases/tag/v0.2.2) (2021-05-03)

### Bug Fixes
- **Pkg.tasker**: DoRun checks if timed out before run (Jitendra Adhikari) [_f27a657_](https://github.com/wct-james/gronx/commit/f27a657)

### Internal Refactors
- **Pkg.tasker**: Use dateFormat var, update final tick phrase (Jitendra Adhikari) [_fad0271_](https://github.com/wct-james/gronx/commit/fad0271)


## [v0.2.1](https://github.com/wct-james/gronx/releases/tag/v0.2.1) (2021-05-02)

### Bug Fixes
- **Pkg.tasker**: Deprecate sleep dur if next tick timeout (Jitendra Adhikari) [_3de45a1_](https://github.com/wct-james/gronx/commit/3de45a1)


## [v0.2.0](https://github.com/wct-james/gronx/releases/tag/v0.2.0) (2021-05-02)

### Features
- **Cmd.tasker**: Add tasker for standalone usage as task daemon (Jitendra Adhikari) [_0d99409_](https://github.com/wct-james/gronx/commit/0d99409)
- **Pkg.tasker**: Add parser for tasker pkg (Jitendra Adhikari) [_e7f1811_](https://github.com/wct-james/gronx/commit/e7f1811)
- **Pkg.tasker**: Add tasker pkg (Jitendra Adhikari) [_a57b1c4_](https://github.com/wct-james/gronx/commit/a57b1c4)

### Bug Fixes
- **Pkg.tasker**: Use log.New() instead (Jitendra Adhikari) [_0cf2c07_](https://github.com/wct-james/gronx/commit/0cf2c07)
- **Validator**: This check is not really required (Jitendra Adhikari) [_c3d75e3_](https://github.com/wct-james/gronx/commit/c3d75e3)

### Internal Refactors
- **Gronx**: Add public methods for internal usage, expose spaceRe (Jitendra Adhikari) [_94eb20b_](https://github.com/wct-james/gronx/commit/94eb20b)

### Miscellaneous
- **Pkg.tasker**: Use file perms as octal (Jitendra Adhikari) [_83f258d_](https://github.com/wct-james/gronx/commit/83f258d)
- **Workflow**: Include all tests in action (Jitendra Adhikari) [_7328cbf_](https://github.com/wct-james/gronx/commit/7328cbf)

### Documentations
- Add task mangager and tasker docs/usages (Jitendra Adhikari) [_e77aa5f_](https://github.com/wct-james/gronx/commit/e77aa5f)


## [v0.1.4](https://github.com/wct-james/gronx/releases/tag/v0.1.4) (2021-04-25)

### Miscellaneous
- **Mod**: 1.13 is okay too (Jitendra Adhikari) [_6c328e7_](https://github.com/wct-james/gronx/commit/6c328e7)
- Try go 1.13.x (Jitendra Adhikari) [_b017ec4_](https://github.com/wct-james/gronx/commit/b017ec4)

### Documentations
- Practical usage (Jitendra Adhikari) [_9572e61_](https://github.com/wct-james/gronx/commit/9572e61)


## [v0.1.3](https://github.com/wct-james/gronx/releases/tag/v0.1.3) (2021-04-22)

### Internal Refactors
- **Checker**: Preserve error, for pos 2 & 4 bail only on due or err (Jitendra Adhikari) [_39a9cd5_](https://github.com/wct-james/gronx/commit/39a9cd5)
- **Validator**: Do not discard error from strconv (Jitendra Adhikari) [_3b0f444_](https://github.com/wct-james/gronx/commit/3b0f444)


## [v0.1.2](https://github.com/wct-james/gronx/releases/tag/v0.1.2) (2021-04-21)

### Features
- Add IsValid() (Jitendra Adhikari) [_150687b_](https://github.com/wct-james/gronx/commit/150687b)

### Documentations
- IsValid usage (Jitendra Adhikari) [_b747116_](https://github.com/wct-james/gronx/commit/b747116)


## [v0.1.1](https://github.com/wct-james/gronx/releases/tag/v0.1.1) (2021-04-21)

### Features
- Add main gronx api (Jitendra Adhikari) [_1b3b108_](https://github.com/wct-james/gronx/commit/1b3b108)
- Add cron segment checker (Jitendra Adhikari) [_a56be7c_](https://github.com/wct-james/gronx/commit/a56be7c)
- Add validator (Jitendra Adhikari) [_455a024_](https://github.com/wct-james/gronx/commit/455a024)

### Miscellaneous
- **Workflow**: Update actions (Jitendra Adhikari) [_8b54cc3_](https://github.com/wct-james/gronx/commit/8b54cc3)
- Init module (Jitendra Adhikari) [_bada37d_](https://github.com/wct-james/gronx/commit/bada37d)
- Add license (Jitendra Adhikari) [_5f20b96_](https://github.com/wct-james/gronx/commit/5f20b96)
- **Gh**: Add meta files (Jitendra Adhikari) [_35a1310_](https://github.com/wct-james/gronx/commit/35a1310)
- **Workflow**: Add lint/test actions (Jitendra Adhikari) [_884d5cb_](https://github.com/wct-james/gronx/commit/884d5cb)
- Add editorconfig (Jitendra Adhikari) [_8b75494_](https://github.com/wct-james/gronx/commit/8b75494)

### Documentations
- On cron expressions (Jitendra Adhikari) [_547fd72_](https://github.com/wct-james/gronx/commit/547fd72)
- Add readme (Jitendra Adhikari) [_3955e88_](https://github.com/wct-james/gronx/commit/3955e88)


next_test.go
package gronx

import (
	"fmt"
	"testing"
	"time"
)

func TestNextTick(t *testing.T) {
	exp := "* * * * *"
	t.Run("next tick incl "+exp, func(t *testing.T) {
		now := time.Now().Format(FullDateFormat)
		next, _ := NextTick(exp, true)
		tick := next.Format(FullDateFormat)
		if now != tick {
			t.Errorf("expected %v, got %v", now, tick)
		}
	})
	t.Run("next tick excl "+exp, func(t *testing.T) {
		expect := time.Now().Add(time.Minute).Truncate(time.Minute).Format(FullDateFormat)
		next, _ := NextTick(exp, false)
		tick := next.Format(FullDateFormat)
		if expect != tick {
			t.Errorf("expected %v, got %v", expect, tick)
		}
	})
}

func TestNextTickAfter(t *testing.T) {
	t.Run("next run after", func(t *testing.T) {
		t.Run("minutes precision", func(t *testing.T) {
			ref, _ := time.Parse(FullDateFormat, "2020-02-02 02:02:00")
			next, err := NextTickAfter("*/5 * * * *", ref, false)
			if err != nil {
				t.Errorf(err.Error())
			}
			if next.Format(FullDateFormat) != "2020-02-02 02:05:00" {
				t.Errorf("2020-02-02 02:02:00 next tick should be 2020-02-02 02:05:00, got %s", next)
			}
		})

		for i, test := range testcases() {
			t.Run(fmt.Sprintf("next run after incl #%d: %s", i, test.Expr), func(t *testing.T) {
				ref, _ := time.Parse(FullDateFormat, test.Ref)
				if next, err := NextTickAfter(test.Expr, ref, true); err == nil {
					actual := next.Format(FullDateFormat)
					if test.Expect != (test.Ref == actual) {
						t.Errorf("[incl] expected %v, got %v", test.Ref, actual)
					}
				}
			})
		}

		gron := New()
		for i, test := range testcases() {
			t.Run(fmt.Sprintf("next run after excl #%d: %s", i, test.Expr), func(t *testing.T) {
				ref, _ := time.Parse(FullDateFormat, test.Ref)
				next, err := NextTickAfter(test.Expr, ref, false)
				if err == nil {
					expect := test.Next
					if expect == "" {
						expect = test.Ref
					}
					actual := next.Format(FullDateFormat)
					if due, _ := gron.IsDue(test.Expr, next); !due {
						t.Errorf("[%s][%s] should be due on %v", test.Expr, test.Ref, next.Format(FullDateFormat))
					}
					if expect != actual {
						t.Errorf("[%s][%s] expected %v, got %v", test.Expr, test.Ref, expect, actual)
					}
				} else {
					fmt.Println(test.Expr+" failed", err)
				}
			})
		}
	})
}


prev_test.go
package gronx

import (
	"fmt"
	"strings"
	"testing"
	"time"
)

func TestPrevTick(t *testing.T) {
	exp := "* * * * *"
	t.Run("prev tick "+exp, func(t *testing.T) {
		ref, _ := time.Parse(FullDateFormat, "2020-02-02 02:02:00")
		prev, _ := PrevTickBefore(exp, ref, true)
		if prev.Format(FullDateFormat) != "2020-02-02 02:02:00" {
			t.Errorf("[incl] expected %v, got %v", ref, prev)
		}

		expect := time.Now().Add(-time.Minute).Truncate(time.Minute).Format(FullDateFormat)
		prev, _ = PrevTick(exp, false)
		if expect != prev.Format(FullDateFormat) {
			t.Errorf("expected %v, got %v", expect, prev)
		}
	})

	t.Run("prev tick excl "+exp, func(t *testing.T) {
		ref, _ := time.Parse(FullDateFormat, "2020-02-02 02:02:00")
		prev, _ := PrevTickBefore(exp, ref, false)
		if prev.Format(FullDateFormat) != "2020-02-02 02:01:00" {
			t.Errorf("[excl] expected %v, got %v", ref, prev)
		}
	})
}

func TestPrevTickBefore(t *testing.T) {
	t.Run("prev tick before", func(t *testing.T) {
		t.Run("minutes precision", func(t *testing.T) {
			ref, _ := time.Parse(FullDateFormat, "2020-02-02 02:02:00")
			next, _ := NextTickAfter("*/5 * * * *", ref, false)
			if next.Format(FullDateFormat) != "2020-02-02 02:05:00" {
				t.Errorf("next tick should be: %s, got: %s", "2020-02-02 02:05:00", next)
			}
			prev, _ := PrevTickBefore("*/5 * * * *", next, false)
			if prev.Format(FullDateFormat) != "2020-02-02 02:00:00" {
				t.Errorf("next > prev should be %s, got %s", "2020-02-02 02:00:00", prev)
			}
		})

		for i, test := range testcases() {
			t.Run(fmt.Sprintf("prev tick #%d: %s", i, test.Expr), func(t *testing.T) {
				ref, _ := time.Parse(FullDateFormat, test.Ref)
				next1, err := NextTickAfter(test.Expr, ref, false)
				if err != nil {
					return
				}

				prev1, err := PrevTickBefore(test.Expr, next1, true)
				if err != nil {
					if strings.HasPrefix(err.Error(), "unreachable year") {
						return
					}
					t.Errorf("%v", err)
				}

				if next1.Format(FullDateFormat) != prev1.Format(FullDateFormat) {
					t.Errorf("next->prev expect %s, got %s", next1, prev1)
				}

				next2, _ := NextTickAfter(test.Expr, next1, false)
				prev2, err := PrevTickBefore(test.Expr, next2, false)
				if err != nil {
					if strings.HasPrefix(err.Error(), "unreachable year") {
						return
					}
					t.Errorf("%s", err)
				}

				if next1.Format(FullDateFormat) != prev2.Format(FullDateFormat) {
					t.Errorf("next->next->prev expect %s, got %s", next1, prev2)
				}
			})
		}
	})
}


batch_test.go
package gronx

import (
	"fmt"
	"testing"
	"time"
)

func TestBatch(t *testing.T) {
	gron := New()

	t.Run("batch no error", func(t *testing.T) {
		ref := time.Now()
		exprs := []string{"* * * * * *", "*  *  *  *  *  *"}
		exprs = append(exprs, fmt.Sprintf("%d * * * * %d", ref.Minute(), ref.Year()))
		exprs = append(exprs, fmt.Sprintf("* * * * * %d-%d", ref.Year()-1, ref.Year()+1))

		for _, expr := range gron.BatchDue(exprs) {
			if expr.Err != nil {
				t.Errorf("%s error: %#v", expr.Expr, expr.Err)
			}
			if !expr.Due {
				t.Errorf("%s must be due", expr.Expr)
			}
		}
	})

	t.Run("batch error", func(t *testing.T) {
		exprs := []string{"* * * *", "A B C D E F"}
		ref, _ := time.Parse(FullDateFormat, "2022-02-02 02:02:02")
		for _, expr := range gron.BatchDue(exprs, ref) {
			if expr.Err == nil {
				t.Errorf("%s expected error", expr.Expr)
			}
			if expr.Due {
				t.Errorf("%s must not be due when there is error", expr.Expr)
			}
		}
	})
}


next.go
package gronx

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// CronDateFormat is Y-m-d H:i (seconds are not significant)
const CronDateFormat = "2006-01-02 15:04"

// FullDateFormat is Y-m-d H:i:s (with seconds)
const FullDateFormat = "2006-01-02 15:04:05"

// NextTick gives next run time from now
func NextTick(expr string, inclRefTime bool) (time.Time, error) {
	return NextTickAfter(expr, time.Now(), inclRefTime)
}

// NextTickAfter gives next run time from the provided time.Time
func NextTickAfter(expr string, start time.Time, inclRefTime bool) (time.Time, error) {
	gron, next := New(), start.Truncate(time.Minute)
	due, err := gron.IsDue(expr, start)
	if err != nil || (due && inclRefTime) {
		return start, err
	}

	segments, _ := Segments(expr)
	if len(segments) > 5 && isUnreachableYear(segments[5], next, inclRefTime, false) {
		return next, fmt.Errorf("unreachable year segment: %s", segments[5])
	}

	next, err = loop(gron, segments, next, inclRefTime, false)
	// Ignore superfluous err
	if err != nil && gron.isDue(expr, next) {
		err = nil
	}
	return next, err
}

func loop(gron Gronx, segments []string, start time.Time, incl bool, reverse bool) (next time.Time, err error) {
	iter, next, bumped := 500, start, false
over:
	for iter > 0 {
		iter--
		for pos, seg := range segments {
			if seg == "*" || seg == "?" {
				continue
			}
			if next, bumped, err = bumpUntilDue(gron.C, seg, pos, next, reverse); bumped {
				goto over
			}
		}
		if !incl && next.Format(FullDateFormat) == start.Format(FullDateFormat) {
			delta := time.Minute
			if reverse {
				delta = -time.Minute
			}
			next, _, err = bumpUntilDue(gron.C, segments[0], 0, next.Add(delta), reverse)
			continue
		}
		return
	}
	return start, errors.New("tried so hard")
}

var dashRe = regexp.MustCompile(`/.*$`)

func isUnreachableYear(year string, ref time.Time, incl bool, reverse bool) bool {
	if year == "*" || year == "?" {
		return false
	}

	edge, inc := ref.Year(), 1
	if !incl {
		if reverse {
			inc = -1
		}
		edge += inc
	}
	for _, offset := range strings.Split(year, ",") {
		if strings.Index(offset, "*/") == 0 || strings.Index(offset, "0/") == 0 {
			return false
		}
		for _, part := range strings.Split(dashRe.ReplaceAllString(offset, ""), "-") {
			val, err := strconv.Atoi(part)
			if err != nil || (!reverse && val >= edge) || (reverse && val < edge) {
				return false
			}
		}
	}
	return true
}

var limit = map[int]int{0: 60, 1: 24, 2: 31, 3: 12, 4: 366, 5: 100}

func bumpUntilDue(c Checker, segment string, pos int, ref time.Time, reverse bool) (time.Time, bool, error) {
	// <minute> <hour> <day> <month> <weekday> <year>
	iter := limit[pos]
	for iter > 0 {
		c.SetRef(ref)
		if ok, _ := c.CheckDue(segment, pos); ok {
			return ref, iter != limit[pos], nil
		}
		ref = bump(ref, pos, reverse)
		iter--
	}
	return ref, false, errors.New("tried so hard")
}

func bump(ref time.Time, pos int, reverse bool) time.Time {
	factor := 1
	if reverse {
		factor = -1
	}

	switch pos {
	case 0:
		ref = ref.Add(time.Duration(factor) * time.Minute)
	case 1:
		ref = ref.Add(time.Duration(factor) * time.Hour)
	case 2, 4:
		ref = ref.AddDate(0, 0, factor)
	case 3:
		ref = ref.AddDate(0, factor, 0)
	case 5:
		ref = ref.AddDate(factor, 0, 0)
	}
	return ref
}


prev.go
package gronx

import (
	"fmt"
	"time"
)

// PrevTick gives previous run time before now
func PrevTick(expr string, inclRefTime bool) (time.Time, error) {
	return PrevTickBefore(expr, time.Now(), inclRefTime)
}

// PrevTickBefore gives previous run time before given reference time
func PrevTickBefore(expr string, start time.Time, inclRefTime bool) (time.Time, error) {
	gron, prev := New(), start.Truncate(time.Minute)
	due, err := gron.IsDue(expr, start)
	if err != nil || (due && inclRefTime) {
		return prev, err
	}

	segments, _ := Segments(expr)
	if len(segments) > 5 && isUnreachableYear(segments[5], prev, inclRefTime, true) {
		return prev, fmt.Errorf("unreachable year segment: %s", segments[5])
	}

	prev, err = loop(gron, segments, prev, inclRefTime, true)
	// Ignore superfluous err
	if err != nil && gron.isDue(expr, prev) {
		err = nil
	}
	return prev, err
}


checker.go
package gronx

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Checker is interface for cron segment due check.
type Checker interface {
	GetRef() time.Time
	SetRef(ref time.Time)
	CheckDue(segment string, pos int) (bool, error)
}

// SegmentChecker is factory implementation of Checker.
type SegmentChecker struct {
	ref time.Time
}

// GetRef returns the current reference time
func (c *SegmentChecker) GetRef() time.Time {
	return c.ref
}

// SetRef sets the reference time for which to check if a cron expression is due.
func (c *SegmentChecker) SetRef(ref time.Time) {
	c.ref = ref
}

// CheckDue checks if the cron segment at given position is due.
// It returns bool or error if any.
func (c *SegmentChecker) CheckDue(segment string, pos int) (due bool, err error) {
	ref, last := c.GetRef(), -1
	val, loc := valueByPos(ref, pos), ref.Location()
	isMonth, isWeekDay := pos == 2, pos == 4

	for _, offset := range strings.Split(segment, ",") {
		mod := (isMonth || isWeekDay) && strings.ContainsAny(offset, "LW#")
		if due, err = c.isOffsetDue(offset, val, pos); due || (!mod && err != nil) {
			return
		}
		if !mod {
			continue
		}
		if last == -1 {
			last = time.Date(ref.Year(), ref.Month(), 1, 0, 0, 0, 0, loc).AddDate(0, 1, 0).Add(-time.Nanosecond).Day()
		}
		if isMonth {
			due, err = isValidMonthDay(offset, last, ref)
		} else if isWeekDay {
			due, err = isValidWeekDay(offset, last, ref)
		}
		if due || err != nil {
			return due, err
		}
	}

	return false, nil
}

func (c *SegmentChecker) isOffsetDue(offset string, val, pos int) (bool, error) {
	if offset == "*" || offset == "?" {
		return true, nil
	}

	bounds, isWeekDay := boundsByPos(pos), pos == 4
	if strings.Contains(offset, "/") {
		return inStep(val, offset, bounds)
	}
	if strings.Contains(offset, "-") {
		if isWeekDay {
			offset = strings.Replace(offset, "7-", "0-", 1)
		}
		return inRange(val, offset, bounds)
	}

	nval, err := strconv.Atoi(offset)
	if err != nil {
		return false, err
	}

	if nval < bounds[0] || nval > bounds[1] {
		return false, fmt.Errorf("segment#%d: '%s' out of bounds(%d, %d)", pos, offset, bounds[0], bounds[1])
	}

	if !isWeekDay && (val == 0 || nval == 0) {
		return nval == 0 && val == 0, nil
	}

	return nval == val || (isWeekDay && nval == 7 && val == 0), nil
}

func valueByPos(ref time.Time, pos int) (val int) {
	switch pos {
	case 0:
		val = ref.Minute()
	case 1:
		val = ref.Hour()
	case 2:
		val = ref.Day()
	case 3:
		val = int(ref.Month())
	case 4:
		val = int(ref.Weekday())
	case 5:
		val = ref.Year()
	}
	return
}

func boundsByPos(pos int) (bounds []int) {
	bounds = []int{0, 0}
	switch pos {
	case 0:
		bounds = []int{0, 59}
	case 1:
		bounds = []int{0, 23}
	case 2:
		bounds = []int{1, 31}
	case 3:
		bounds = []int{1, 12}
	case 4:
		bounds = []int{0, 7}
	case 5:
		bounds = []int{1, 9999}
	}
	return
}


cmd/tasker/main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/wct-james/gronx/pkg/tasker"
)

var exit = os.Exit
var tick = time.Minute

var opt tasker.Option
var v bool

// Version of tasker, injected in build
var Version = "n/a"

func init() {
	flag.StringVar(&opt.File, "file", "", "The task file in crontab format (without user)")
	flag.StringVar(&opt.Tz, "tz", "Local", "The timezone to use for tasks")
	flag.StringVar(&opt.Shell, "shell", tasker.Shell()[0], "The shell to use for running tasks")
	flag.StringVar(&opt.Out, "out", "", "The fullpath to file where output from tasks are sent to")
	flag.BoolVar(&opt.Verbose, "verbose", false, "The verbose mode outputs as much as possible")
	flag.Int64Var(&opt.Until, "until", 0, "The timeout for task daemon in minutes")
	flag.BoolVar(&v, "v", false, "Show version")
}

func main() {
	mustParseOption()

	taskr := tasker.New(opt)
	for _, task := range tasker.MustParseTaskfile(opt) {
		taskr.Task(task.Expr, taskr.Taskify(task.Cmd, opt))
	}

	if opt.Until > 0 {
		taskr.Until(time.Duration(opt.Until) * tick)
	}

	taskr.Run()
}

func mustParseOption() {
	opt = tasker.Option{}
	flag.Parse()

	if v {
		fmt.Printf("v%s\n", Version)
		exit(0)
	}

	if opt.File == "" {
		flag.Usage()
		exit(1)
	}

	if _, err := os.Stat(opt.File); err != nil {
		log.Printf("can't read taskfile: %s", opt.File)
		exit(1)
	}
}


gronx.go
package gronx

import (
	"errors"
	"regexp"
	"strings"
	"time"
)

var literals = strings.NewReplacer(
	"SUN", "0", "MON", "1", "TUE", "2", "WED", "3", "THU", "4", "FRI", "5", "SAT", "6",
	"JAN", "1", "FEB", "2", "MAR", "3", "APR", "4", "MAY", "5", "JUN", "6", "JUL", "7",
	"AUG", "8", "SEP", "9", "OCT", "10", "NOV", "11", "DEC", "12",
)

var expressions = map[string]string{
	"@yearly":    "0 0 1 1 *",
	"@annually":  "0 0 1 1 *",
	"@monthly":   "0 0 1 * *",
	"@weekly":    "0 0 * * 0",
	"@daily":     "0 0 * * *",
	"@hourly":    "0 * * * *",
	"@always":    "* * * * *",
	"@5minutes":  "*/5 * * * *",
	"@10minutes": "*/10 * * * *",
	"@15minutes": "*/15 * * * *",
	"@30minutes": "0,30 * * * *",
}

// SpaceRe is regex for whitespace.
var SpaceRe = regexp.MustCompile(`\s+`)

func normalize(expr string) []string {
	expr = strings.Trim(expr, " \t")
	if e, ok := expressions[strings.ToLower(expr)]; ok {
		expr = e
	}

	expr = SpaceRe.ReplaceAllString(expr, " ")
	expr = literals.Replace(strings.ToUpper(expr))

	return strings.Split(strings.ReplaceAll(expr, "  ", " "), " ")
}

// Gronx is the main program.
type Gronx struct {
	C Checker
}

// New initializes Gronx with factory defaults.
func New() Gronx {
	return Gronx{&SegmentChecker{}}
}

// IsDue checks if cron expression is due for given reference time (or now).
// It returns bool or error if any.
func (g *Gronx) IsDue(expr string, ref ...time.Time) (bool, error) {
	ref = append(ref, time.Now())
	g.C.SetRef(ref[0])

	segs, err := Segments(expr)
	if err != nil {
		return false, err
	}

	return g.SegmentsDue(segs)
}

func (g *Gronx) isDue(expr string, ref time.Time) bool {
	due, err := g.IsDue(expr, ref)
	return err == nil && due
}

// Segments splits expr into array of cron parts.
// It returns array or error.
func Segments(expr string) ([]string, error) {
	segs := normalize(expr)
	slen := len(segs)
	if slen < 5 || slen > 6 {
		return []string{}, errors.New("expr should contain 5-6 segments separated by space")
	}

	return segs, nil
}

// SegmentsDue checks if all cron parts are due.
// It returns bool. You should use IsDue(expr) instead.
func (g *Gronx) SegmentsDue(segs []string) (bool, error) {
	for pos, seg := range segs {
		if seg == "*" || seg == "?" {
			continue
		}

		if due, err := g.C.CheckDue(seg, pos); !due {
			return due, err
		}
	}

	return true, nil
}

// IsValid checks if cron expression is valid.
// It returns bool.
func (g *Gronx) IsValid(expr string) bool {
	segs, err := Segments(expr)
	if err != nil {
		return false
	}

	g.C.SetRef(time.Now())
	for pos, seg := range segs {
		if _, err := g.C.CheckDue(seg, pos); err != nil {
			return false
		}
	}

	return true
}
